### PR TITLE
Porting combined errors from mbedtls changes

### PR DESCRIPTION
**Pull request by mridul-manohar**
Due to downgrade back to v2.28 (from ~3.X), the enhancement in [PR #271](https://github.com/fortanix/rust-mbedtls/pull/271/files)  was lost in the mbedtls 2.8 branch being used. 
we need to port these combined errors from mbedtls changes back to the mbedtls 2.8 branch and build with the latest upgrade in rust toolchain version to apply the enhancement.
This PR ports said changes to a latest branch forked from mbedtls master and builds the same on latest rustc 1.83.0-nightly (26d8e9255 2024-10-11) version.
https://fortanix.atlassian.net/browse/PROD-9297

### CODE REVIEW

File: mbedtls/src/bignum/mod.rs [hunk 1]
```diff
@@ -6,7 +6,10 @@
  * option. This file may not be copied, modified, or distributed except
  * according to those terms. */
 
-use crate::error::{Error, IntoResult, Result};
+
+ #[cfg(feature = "std")]
```
Comment (Taowyoo) [inline mbedtls/src/bignum/mod.rs:10]: extra leading space.

File: mbedtls/src/pk/mod.rs [hunk 2]
```diff
@@ -1543,7 +1543,7 @@ iy6KC991zzvaWY/Ys+q/84Afqa+0qJKQnPuy/7F5GkVdQA/lfbhi
         });
         assert_eq!(
             pk.verify(digest, data, &signature[0..len]).unwrap_err(),
-            Error::RsaInvalidPadding
+            codes::RsaInvalidPadding.into()
         );
     }
```
Comment (Taowyoo) [inline mbedtls/src/pk/mod.rs:1549]: The error types in tests are not updated correctly according to the failed CI. Please update them

File: mbedtls/Cargo.toml
Comment (Taowyoo) [inline mbedtls/Cargo.toml:1]: Please bump the `mbedtls` version to `0.13.0` since changing error type is a breaking change.

**Review by Taowyoo — COMMENTED** [review]
I saw some unaligned spaces. And **Rustfmt** CI is failed. Please run
```shell
  cargo fmt -p mbedtls
  cargo fmt -p mbedtls-platform-support
  cargo fmt -p mbedtls-sys-auto
```
to ensure everything is formated

File: mbedtls/src/bignum/mod.rs (hunk 1 above)
Comment (mridul-manohar) [reply mbedtls/src/bignum/mod.rs]: removed extra space

File: mbedtls/Cargo.toml
Comment (mridul-manohar) [reply mbedtls/Cargo.toml:1]: updated `mbedtls` version to `0.13.0`

File: mbedtls/src/cipher/raw/mod.rs [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 use mbedtls_sys::*;
 
-use crate::error::{Error, IntoResult, Result};
+use crate::error::{codes, IntoResult, Result};
```
Comment (Taowyoo) [inline mbedtls/src/cipher/raw/mod.rs:11]: just a reminder: 
need to ensure there is no usage of `Error` in this file

File: mbedtls/src/hash/mod.rs [hunk 4]
```diff
@@ -6,7 +6,7 @@
  * option. This file may not be copied, modified, or distributed except
  * according to those terms. */
 
-use crate::error::{Error, IntoResult, Result};
+use crate::error::{codes, IntoResult, Result};
```
Comment (Taowyoo) [inline mbedtls/src/hash/mod.rs:9]: just a reminder:
need to ensure there is no usage of `Error` in this file

File: mbedtls/src/pk/dsa/mod.rs [hunk 5]
```diff
@@ -9,9 +9,11 @@
 use crate::bignum::Mpi;
 use crate::hash::{MdInfo, Type as MdType};
 use crate::pk::rfc6979::generate_rfc6979_nonce;
-use crate::rng::Random;
-use crate::{Error, Result};
+#[cfg(not(feature = "std"))]
```
Comment (Taowyoo) [inline mbedtls/src/pk/dsa/mod.rs:12]:
```suggestion
```
i think this line is not necessary.

File: mbedtls/src/ssl/config.rs [hunk 6]
```diff
@@ -20,7 +20,8 @@ use mbedtls_sys::*;
 use crate::alloc::List as MbedtlsList;
 #[cfg(not(feature = "std"))]
 use crate::alloc_prelude::*;
-use crate::error::{Error, IntoResult, Result};
+#[cfg(feature = "std")]
```
Comment (Taowyoo) [inline mbedtls/src/ssl/config.rs:23]: Similar, if it's necessary please ignore this
```suggestion
```

File: mbedtls/src/pk/mod.rs (hunk 2 above)
Comment (mridul-manohar) [reply mbedtls/src/pk/mod.rs:1552]: addressed this comment.

File: mbedtls/src/cipher/raw/mod.rs (hunk 3 above)
Comment (mridul-manohar) [reply mbedtls/src/cipher/raw/mod.rs:11]: done.

File: mbedtls/src/hash/mod.rs (hunk 4 above)
Comment (mridul-manohar) [reply mbedtls/src/hash/mod.rs:9]: done.

File: mbedtls/src/pk/dsa/mod.rs (hunk 5 above)
Comment (mridul-manohar) [reply mbedtls/src/pk/dsa/mod.rs]: removed std config.

**Review by Taowyoo — COMMENTED** [review]
LGTM,

File: mbedtls/src/error.rs [hunk 7]
```diff
@@ -23,66 +24,114 @@ pub trait IntoResult: Sized {
...
-                let err_code = match self {
-                    _ if self >= 0 => return Ok(self),
-                    ERR_UTF8_INVALID => return Err(Error::Utf8Error(None)),
-                    _ => -self,
-                };
-                let (high_level_code, low_level_code) = (err_code & 0xFF80, err_code & 0x7F);
-                Err($n::from_mbedtls_code(if high_level_code > 0 { -high_level_code } else { -low_level_code }))
+        impl From<c_int> for $n {
+            fn from(code: c_int) -> $n {
+                // check against mask here (not in match blook) to make it compile-time
```
Comment (zugzwang) [inline mbedtls/src/error.rs:52]: nit: `blook` -> `block`

Comment (zugzwang) [reply mbedtls/src/error.rs:52]: (I understand that this comes from the original PR, though)

File: mbedtls/src/error.rs [hunk 8]
```diff
@@ -23,66 +24,114 @@ pub trait IntoResult: Sized {
...
 // circumstances (callbacks from mbedtls to rust) do we need to pass a Utf8Error
 // back in to mbedtls.
 pub const ERR_UTF8_INVALID: c_int = -0x10000;
 
 macro_rules! error_enum {
-    {enum $n:ident {$($rust:ident = $c:ident,)*}} => {
-        #[derive(Debug, Eq, PartialEq)]
+    {
+        const MASK: c_int = $mask:literal;
+        enum $n:ident {$($rust:ident = $c:ident,)*}
```
Comment (zugzwang) [inline mbedtls/src/error.rs:41]: nit: `$n` could be called `$error` or something else. 

(This is also part of the original PR, we can just ignore this but I just wanted to object about it)

File: mbedtls/src/lib.rs [hunk 9]
```diff
@@ -27,7 +27,7 @@ mod wrapper_macros;
 //      API
 // ==============
 pub mod bignum;
-mod error;
+pub mod error;
```
Comment (zugzwang) [inline mbedtls/src/lib.rs:30]: I'm not in favor of this module export, but also part of the original PR

**Review by zugzwang — APPROVED** [review]
Thanks for the PR

File: mbedtls/src/error.rs (hunk 7 above)
Comment (mridul-manohar) [reply mbedtls/src/error.rs:52]: addressed this. `blook -> block`

File: mbedtls/src/error.rs (hunk 8 above)
Comment (mridul-manohar) [reply mbedtls/src/error.rs:41]: addressed this. `$n -> $error`

**Review by Taowyoo — APPROVED** [review]

**Pull request merged**